applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h

```diff
@@ -221,6 +226,37 @@ class GeometryUtilities
     	KRATOS_CATCH("");
     }
 
+    // --------------------------------------------------------------------------
+    void ExtractEdgeNodes( std::string const& rEdgeSubModelPartName ) {
+        KRATOS_TRY;
+
+        KRATOS_ERROR_IF(mrModelPart.Elements().size() == 0) << "ExtractEdgeNodes: No elements defined. Automatic edge detection will not find any edge nodes!" << std::endl;
+
+        ModelPart& r_edge_model_part = mrModelPart.GetSubModelPart(rEdgeSubModelPartName);
+
+        KRATOS_ERROR_IF(r_edge_model_part.Nodes().size() != 0) << "ExtractEdgeNodes: The edge model part already has nodes!" << std::endl;
+
+        for (auto& r_node_i : mrModelPart.Nodes()) {
+            Node<3>* p_node_i = &r_node_i;
+            GlobalPointersVector< Node<3 > >& r_node_i_neighbours = p_node_i->GetValue(NEIGHBOUR_NODES);
+            for( GlobalPointersVector< Node<3> >::iterator j =	r_node_i_neighbours.begin(); j != r_node_i_neighbours.end(); j++) {
+                GlobalPointersVector<Element>& r_element_neighbours = p_node_i->GetValue(NEIGHBOUR_ELEMENTS);
+                int count = 0;
+                for( GlobalPointersVector<Element>::iterator k = r_element_neighbours.begin(); k != r_element_neighbours.end(); k++) {
```

Comment:
```suggestion
                for(auto k = r_element_neighbours.begin(); k != r_element_neighbours.end(); ++k) {
```

Reply:
Can I even use `auto` for the `GlobalPointersVector`s, i.e. `GlobalPointersVector<Node<3>>` and `GlobalPointersVector<Element>`, inside the loops as well?

Reply:
sure!
Just give it a try, the compiler will complain right away :)